experiment/selinux/src/selinux.rs

```diff
@@ -44,268 +95,301 @@ impl SELinux {
     pub fn new() -> Self {
         SELinux {
             have_thread_self: AtomicBool::new(false),
-            init_done: AtomicBool::new(false),
-        }
-    }
+            attr_path_init_done: AtomicBool::new(false),
 
-    // function similar with setDisabled in go-selinux repo.
-    // set_disabled disables SELinux support for the package.
-    pub fn set_disabled() {
-        unimplemented!("not implemented yet")
-    }
+            selinuxfs_init_done: AtomicBool::new(false),
+            selinuxfs: None,
 
-    // function similar with getEnabled in go-selinux repo.
-    // get_enabled returns whether SELinux is enabled or not.
-    pub fn get_enabled() -> bool {
-        unimplemented!("not implemented yet")
-    }
+            policy_root_init_done: AtomicBool::new(false),
+            policy_root: None,
 
-    // function similar with classIndex in go-selinux repo.
-    // classIndex returns the int index for an object class in the loaded policy,
-    // or -1 and an error.
-    pub fn class_index(class: &str) -> Result<i64, String> {
-        unimplemented!("not implemented yet")
-    }
+            load_labels_init_done: AtomicBool::new(false),
+            labels: HashMap::new(),
 
-    // function similar with setFileLabel in go-selinux repo.
-    // set_file_label sets the SELinux label for this path, following symlinks, or returns an error.
-    pub fn set_file_label(fpath: &Path, label: &str) -> Result<(), SELinuxError> {
-        if !fpath.exists() {
-            return Err(SELinuxError::SetFileLabel(ERR_EMPTY_PATH.to_string()));
+            read_only_file_label: None,
         }
+    }
 
-        loop {
-            match fpath.set_xattr(XATTR_NAME_SELINUX, label.as_bytes()) {
-                Ok(_) => break,
-                // When a system call is interrupted by a signal, it needs to be retried.
-                Err(XattrError::EINTR(_)) => continue,
-                Err(e) => {
-                    return Err(SELinuxError::SetFileLabel(format!(
-                        "set_xattr failed: {}",
-                        e
-                    )));
+    // This function returns policy_root.
+    // Directories under policy root has configuration files etc.
+    fn policy_root(&mut self) -> Option<&PathBuf> {
+        // Avoiding code conflicts and ensuring thread-safe execution once only.
+        if !self.policy_root_init_done.load(Ordering::SeqCst) {
+            let policy_root_path = Self::read_config(SELINUX_TYPE_TAG).unwrap_or_default();
+            self.policy_root = Some(PathBuf::from(policy_root_path));
+            self.policy_root_init_done.store(true, Ordering::SeqCst);
+        }
+        self.policy_root.as_ref()
+    }
+
+    // This function reads SELinux config file and returns the value with a specified key.
+    fn read_config(target: &str) -> Result<String, SELinuxError> {
```

Comment:
What is `target`? Can we use a type instead of  `&str`?

Reply:
`target` means a specified key in selinux config file.
In this cargo, read_config is used finding values of `SELINUX` and `SELINUXTYPE`.
However it is possible that there are other keys used in selinux config file, therefore I am not using a type instead of `&str`.